autogen/coding/embedded_ipython_code_executor.py

```diff
@@ -0,0 +1,221 @@
+
+    **This will execute LLM generated code on the local machine.**
+
+    Each execution is stateful and can access variables created from previous
+    executions in the same session. The kernel must be installed before using
+    this class. The kernel can be installed using the following command:
+    `python -m ipykernel install --user --name {kernel_name}`
+    where `kernel_name` is the name of the kernel to install.
+
+    Args:
+        timeout (int): The timeout for code execution, by default 60.
+        kernel_name (str): The kernel name to use. Make sure it is installed.
+            By default, it is "python3".
+        output_dir (str): The directory to save output files, by default ".".
+    """
+
+    class UserCapability:
+        """An AgentCapability class that gives agent ability use a stateful
+        code executor."""
+
+        DEFAULT_SYSTEM_MESSAGE_UPDATE = """# IPython Coding Capability
+You have been given coding capability to solve tasks using Python code in a stateful IPython kernel.
+You are responsible for writing the code, and the user is responsible for executing the code.
+
+When you write Python code, put the code in a markdown code block with the language set to Python.
+For example:
+```python
+x = 3
+```
+You can use the variable `x` in subsequent code blocks.
+```python
+print(x)
+```
+
+Write code incrementally and leverage the statefulness of the kernel to avoid repeating code.
+Import libraries in a separate code block.
+Define a function or a class in a separate code block.
+Run code that produces output in a separate code block.
+Run code that involves expensive operations like download, upload, and call external APIs in a separate code block.
+
+When your code produces an output, the output will be returned to you.
+Because you have limited conversation memory, if your code creates an image,
+the output will be a path to the image instead of the image itself.
+"""
+
+        def add_to_agent(self, agent: LLMAgent) -> None:
+            """Add this capability to an agent."""
+            # system message is a string or a list of strings
+            system_message = agent.system_message + self.DEFAULT_SYSTEM_MESSAGE_UPDATE
+            agent.update_system_message(system_message)
+
+    timeout: int = Field(default=60, ge=1, description="The timeout for code execution.")
+    kernel_name: str = Field(default="python3", description="The kernel name to use. Make sure it is installed.")
+    output_dir: str = Field(default=".", description="The directory to save output files.")
+
+    @validator("output_dir")
+    def _output_dir_must_exist(cls, value: str) -> str:
+        if not os.path.exists(value):
+            raise ValueError(f"Output directory {value} does not exist.")
+        return value
+
+    def __init__(self, **kwargs: Any):
+        super().__init__(**kwargs)
+        # Check if the kernel is installed.
+        if self.kernel_name not in KernelSpecManager().find_kernel_specs():
```

Comment:
The kernel name and kernel spec are different things aren't they?

I could be wrong but i thought you isntantiate a kernel from a spec

Reply:
Correct kernel name and kernel specs are different. `find_kernel_specs()` returns a dictionary mapping kernel name to kernel specs. So we can use this here to check whether the kernel name is available. 

https://jupyter-client.readthedocs.io/en/5.2.2/api/kernelspec.html#jupyter_client.kernelspec.KernelSpecManager.find_kernel_specs

Reply:
@jackgerrits would you like to resolve this one? Currently I am starting a kernel based on the name.